source/tit/core/stats.hpp

```diff
@@ -0,0 +1,225 @@
+  auto operator=(BaseStatsVar&&) -> BaseStatsVar& = delete;
+
+  /// Statistics variable is not copyable.
+  auto operator=(const BaseStatsVar&) -> BaseStatsVar& = delete;
+
+  /// Destroy the statistics variable.
+  virtual ~BaseStatsVar() = default;
+
+  /// Get the average value as a string.
+  constexpr virtual auto render_avg() const -> std::string = 0;
+
+  /// Get the minimum value as a string.
+  constexpr virtual auto render_min() const -> std::string = 0;
+
+  /// Get the maximum value as a string.
+  constexpr virtual auto render_max() const -> std::string = 0;
+
+}; // class BaseStatsVar
+
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+/// Valid underlying types for statistics variables.
+template<class Val>
+concept stattable = //
+    std::is_object_v<Val> && std::semiregular<Val> &&
+    std::formattable<Val, char> && requires(Val val, size_t n) {
+      // Can compute the average value.
+      { val += val } -> std::convertible_to<Val&>;
+      { val / n } -> std::convertible_to<Val>;
+      // Can compute the minimum and maximum values.
+      { std::min(val, val) } -> std::convertible_to<Val>;
+      { std::max(val, val) } -> std::convertible_to<Val>;
+    };
+
+/// Statistics variable.
+template<class Val>
+class StatsVar;
+
+template<stattable Val>
+class StatsVar<Val> final : public BaseStatsVar {
+public:
+
+  /// Get the mean value as a string.
+  constexpr auto render_avg() const -> std::string override {
+    return std::format("{}", sum_ / count_);
+  }
+
+  /// Get the minimum value as a string.
+  constexpr auto render_min() const -> std::string override {
+    return std::format("{}", min_);
+  }
+
+  /// Get the maximum value as a string.
+  constexpr auto render_max() const -> std::string override {
+    return std::format("{}", max_);
+  }
+
+  /// Update the statistics variable.
+  void update(const Val& val) {
+    using std::max;
+    using std::min;
+    if (count_ == 0) sum_ = min_ = max_ = val;
+    else sum_ += val, min_ = min(min_, val), max_ = max(max_, val);
+    count_ += 1;
+  }
+
+private:
+
+  size_t count_ = 0;
+  Val sum_{};
+  Val min_{};
+  Val max_{};
+
+}; // class StatsVar
+
+template<std::ranges::input_range Vals>
+  requires stattable<std::ranges::range_value_t<Vals>>
+class StatsVar<Vals> final : public BaseStatsVar {
+public:
+
+  /// Value type.
+  using Val = std::ranges::range_value_t<Vals>;
+
+  /// Construct the statistics variable.
+  StatsVar() = default;
+
+  /// Get the mean value as a string.
+  constexpr auto render_avg() const -> std::string override {
+    auto result = std::format("{}", sum_.front() / count_);
+    for (const auto& val : sum_ | std::views::drop(1)) {
+      result += std::format(" {}", val / count_);
+    }
+    return result;
+  }
+
+  /// Get the minimum value as a string.
+  constexpr auto render_min() const -> std::string override {
+    auto result = std::format("{}", min_.front());
```

Comment:
Let's introduce a helper function for formatting ranges until we get the range formatting support by the compiler.